index.ts

```diff
@@ -1,35 +1,65 @@
 export class OatyArray {
-  private _original: object[]
-  private _keys: string[]
-  private _transposed: object = {}
+  private keys: string[] | undefined
+  private original: object[] = []
+  private transposed: { [key: string]: any } = {}
+
+  constructor(options?: { data?: object[], keys?: string[] }) {
+    if (options !== undefined) {
+      if (options.data !== undefined) { this.push(...options.data) }
+      if (options.keys !== undefined) { this.keys = options.keys }
+    }
+  }
 
   get length(): number {
-    return this._original.length
+    return this.original.length
   }
 
-  constructor(input: object[], keys: string[]) {
-    this._original = input.map((x) => Object.assign({}, x))
-    this._keys = keys
+  public getKeys(): string[] | undefined {
+    return this.keys
+  }
 
-    this._transposed = transpose(this._keys, this._original)
+  public getOriginal(): object[] {
+    return this.original
   }
 
-  public get(keyName: string, keyValue: string): object[] {
-    return this._transposed[keyName][keyValue]
+  public getTransposed(): object[] {
+    return this.original
   }
 
-  public push(...items: object[]): number {
-    this._transposed = transpose(this._keys, items, this._transposed)
-    return this._original.push(...items)
+  public get(keyName: string, keyValue: string): object[] | undefined {
+    if (this.transposed === undefined
+     || this.transposed[keyName] === undefined) {
+      return undefined
+    }
+    return this.transposed[keyName][keyValue]
   }
-}
 
-const transpose = (keys: string[], items: object[], current?: object) => {
-  return keys.reduce((transposed, key) => {
-    transposed[key] = items.reduce((acc, item) => {
-      acc[item[key]] ? acc[item[key]].push(item) : acc[item[key]] = [item]
-      return acc
-    }, current || {})
-    return transposed
-  }, {})
+  public push(...data: object[]) {
+    data.forEach((datum) => {
+      const clone = Object.assign({}, datum)
+      this.original.push(clone)
+      this.transpose(clone)
+    })
+    return this.length
+  }
+
+  private getKeysForObject(datum: object): string[] {
+    return this.keys !== undefined ? this.keys : Object.keys(datum)
+  }
+
+  private transpose(datum: { [key: string]: any }) {
+    this.getKeysForObject(datum).forEach((key) => {
+      if (datum[key] === undefined) {
+        return
+      }
+      if (this.transposed[key] === undefined) {

```

Comment:
lots of `undefined` checks in this method that weren't needed before - why now?

Reply:
I wanted to just do `this.transposed[key][datum[key]] = [datum]` but it throws if `this.transposed[key]` is undefined. 
Same as another comment: If options.keys is defined then we want to transpose only those keys, otherwise we'll use the object's.

Reply:
imo, `=== undefined` is rarely needed, but this relatively short function now uses it 3 times

Reply:
Is there a better way to `this.transposed[key][datum[key]].push(datum)` given `this.transposed[key]` or `this.transposed[key][datum[key]]` could be undefined?

Reply:
Continued in next review